### PR TITLE
Allow states on sub-models #4318

### PR DESCRIPTION
**Pull request by cybercop23**
Allows adding states on submodels, so you can have both faces on main and states on sub-models

### CODE REVIEW

**Comment by cybercop23** [comment]
Correctamundo... that's why you'd use/create a full/virtual model that contains all nodes in the model and place it on that.

**Comment by computergeek1507** [comment]
The states also have to be defined to match the Submodels nodes, not the model, which would show up incorrectly in the states dialog. Submodel of 10-20 who be 1-10 in the state dialog

**Comment by cybercop23** [comment]
I think Scott and I are saying the same thing, right @computergeek1507? (just want to make sure)

**Comment by cybercop23** [comment]
Well... happy to test other cases, but the one that we developed for and tested in zoom worked as expected. Granted only some advanced users will use this to "blend" faces and states, but it eliminates having groups for submodels and allows you to just place it right on the model.

**Comment by cybercop23** [comment]
So that the States are defied based on the "full" model so you don't have the 10-20 => 1-10.

**Comment by cybercop23** [comment]
Exactly, so that you can easily sequence on the model and only one model. Seems a heck of a lot easier than shadow model (or groups) to place face and states effects.

**Comment by cybercop23** [comment]
b/c of the way states black out certain nodes on diff timings. just tried it and with layers and faces at the top then states, or vice a versa doesn't give the effect we want. only once you move the state to the bottom on the shadow/virtual sub-model

**Comment by cybercop23** [comment]
Good point. Just verified... no crash.. static (as expected) since the states nodes are out of bounds.

**Comment by cybercop23** [comment]
OHHH MAN... that was my initial dev path. Add one to the bottom of the list for each model, that's in essence the first one [0] that's the model itself. I can work on that ... so long it gets pre-approved since I spent quite some time and wasn't getting it... :)

**Comment by cybercop23** [comment]
> I can see that being a pain though cause then every time InitModel runs its going to have to access whether the submodel has been created already.

dedup list?

File: xLights/effects/StateEffect.cpp [hunk 1]
```diff
@@ -234,7 +234,11 @@ void StateEffect::RenderState(RenderBuffer &buffer,
 
     Model* model_info = buffer.GetModel();
     if (model_info == nullptr) {
-        return;
+        buffer.cur_model = buffer.cur_model.substr(0, buffer.cur_model.find("/"));
```
Comment (derwin12) [inline xLights/effects/StateEffect.cpp:237]: (I don't know) ... but do you need to make sure buffer != nullptr?

Comment (cybercop23) [reply xLights/effects/StateEffect.cpp:237]: not in this case. you hit this when ytou place a state effect on the sub-model (only); not hitting this if you do it at the model level; by then the buffer is alreay avail.

**Review by computergeek1507 — APPROVED** [review]

**Pull request merged**